cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,458 @@
+}
+
+func validateInstance(validatedParams *validatedParams, sourceOS string) error {
+	inst, err := computeClient.GetInstance(validatedParams.project, validatedParams.zone, validatedParams.instanceName)
+	if err != nil {
+		return daisy.Errf("Failed to get instance: %v", err)
+	}
+	if err := validateLicense(inst, sourceOS); err != nil {
+		return err
+	}
+
+	if err := validateOSDisk(inst.Disks[0], validatedParams, err); err != nil {
+		return err
+	}
+
+	for _, metadataItem := range inst.Metadata.Items {
+		if metadataItem.Key == metadataKeyWindowsStartupScriptURL {
+			windowsStartupScriptURLBackup = metadataItem.Value
+		} else if metadataItem.Key == metadataKeyWindowsStartupScriptURLBackup {
+			windowsStartupScriptURLBackupExists = true
+		}
+	}
+	// If script url backup exists, don't backup again to overwrite it
+	if windowsStartupScriptURLBackupExists {
+		windowsStartupScriptURLBackup = nil
+		fmt.Printf("\n'%v' was backed up to '%v' before.\n\n",
+			metadataKeyWindowsStartupScriptURL, metadataKeyWindowsStartupScriptURLBackup)
+	}
+	return nil
+}
+
+func validateOSDisk(osDisk *compute.AttachedDisk, validatedParams *validatedParams, err error) error {
+	validatedParams.osDisk = param.GetZonalResourcePath(validatedParams.zone, "disks", osDisk.Source)
+	osDiskName := getResourceRealName(osDisk.Source)
+	d, err := computeClient.GetDisk(validatedParams.project, validatedParams.zone, osDiskName)
+	if err != nil {
+		return daisy.Errf("Failed to get OS disk info: %v", err)
+	}
+	validatedParams.osDiskDeviceName = osDisk.DeviceName
+	validatedParams.osDiskAutoDelete = osDisk.AutoDelete
+	validatedParams.osDiskType = getResourceRealName(d.Type)
+	return nil
+}
+
+func validateLicense(inst *compute.Instance, sourceOS string) error {
+	matchSourceOSVersion := false
+	upgraded := false
+	if len(inst.Disks) == 0 {
+		return daisy.Errf("No disks attached to the instance.")
+	}
+	for _, lic := range inst.Disks[0].Licenses {
+		if strings.HasSuffix(lic, expectedLicense[sourceOS]) {
+			matchSourceOSVersion = true
+		} else if strings.HasSuffix(lic, appendLicense[sourceOS]) {
+			upgraded = true
+		}
+	}
+	if !matchSourceOSVersion {
+		return daisy.Errf(fmt.Sprintf("Can only upgrade GCE instance with %v license attached", expectedLicense[sourceOS]))
+	}
+	if upgraded {
+		return daisy.Errf(fmt.Sprintf("The GCE instance is with %v license attached, which measn it either has been upgraded or has started a upgrade in the past.", appendLicense[sourceOS]))
+	}
+	return nil
+}
+
+func runUpgradeWorkflow(ctx context.Context, currentExecutablePath string,
+	skipMachineImageBackup bool, autoRollback bool, sourceOS string, targetOS string, instance string,
+	project string, zone string, instanceName string, oldOSDisk string, osDiskType string,
+	osDiskDeviceName string, osDiskAutoDelete bool, timeout string, scratchBucketGcsPath string, oauth string,
+	ce string, gcsLogsDisabled bool, cloudLogsDisabled bool, stdoutLogsDisabled bool) (*daisy.Workflow, error) {
```

Comment:
I've found that functions with large arity are difficult to read and maintain. Some specifics:

1. When calling the function, it's easy to swap parameters. The compiler can help if you have the wrong type, but  when there are many instances of the same type (string, string, string, string.....).
2. I see both `instance string` and `instanceName string` -- without reading the function, I don't know how those are different.
3. It encourages future people to continue adding parameters. (Take a look at the parameters I've added to importer.Run!)

Some ideas for avoiding this:
1. Break the function into smaller parts.
1. Use a receiver object
1. Wrap primitive types in domain-specific types. (instead of `oldOSDisk string`, use `oldOS Disk`)
1. Combine several related arguments into a domain-specific object . This would hep with  `instance` and `instanceName`.

Reply:
changed, please review the latest version